### PR TITLE
Decomposition for OBBT

### PR DESCRIPTION
**Pull request by michaelbynum**
This PR adds some tools (functions and classes) for performing decomposed bounds tightening (DBT). The main functions meant to be used by users `decompose_model` and `perform_dbt`. The function `decompose_mdoel` takes a Pyomo model and returns a `TreeBlock`, which is a derived Pyomo block containing all of the information needed to perform DBT. A `TreeBlock` has children blocks and linking constraints. The children blocks are also `TreeBlock`s. Each `TreeBlock` is a leaf if and only if it does not have any children blocks. The linking constraints link the variables from children blocks which were partitioned on by Metis. A `TreeBlock` also has methods to get the number of stages, iterate over the blocks in a particular stage, and, given a block, get the stage that block belongs to. The `perform_dbt` method takes in a `TreeBlock` and performs OBBT according to the structure of the `TreeBlock`. There is also a function called `perform_dbt_with_integers_relaxed` which has special handling for binary and integer variables. 

The primary file is Coramin/domain_reduction/dbt.py

### CODE REVIEW

**Comment by carldlaird** [comment]
Michael, can you add short description of this that explains a little more than the title?

File: coramin/domain_reduction/tests/test_decomp.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
+from coramin.domain_reduction.dbt import TreeBlock, TreeBlockError, convert_pyomo_model_to_bipartite_graph, \
```
Comment (carldlaird) [inline coramin/domain_reduction/tests/test_decomp.py]: This is minor and should not hold up the PR, but can we change the name of this file to test_dbt? "decomp" is really general and can refer to many things. dbt is the shortened named for decomposed_bounds_tightening.

**Review by carldlaird — APPROVED** [review]
This seems like reasonable, standalone functionality that should not impact other parts of the code. Please consider the changes asked for in the conversation and as part of this review.

Inline documentation on usage could be improved.

File: .github/workflows/main_ci.yml [hunk 2]
```diff
@@ -24,8 +24,10 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install pyomo numpy scipy nose codecov coverage cplex==12.9
-        conda install mpi4py ipopt glpk
+        pip install pyomo numpy scipy nose codecov coverage cplex==12.9 sympy
+        conda install mpi4py metis ipopt glpk wntr networkx=2.3 --no-update-deps
```
Comment (jsiirola) [inline .github/workflows/main_ci.yml:31]: Why the hard-coded networkx and cplex versions?  Can you add a comment as to why this was pinned and what needs to be resolved to free the dependency.

**Comment by bknueven** [comment]
My only comment is regarding integer variables--when working on the integer relaxation for bounds tightening it seems we don't exploit the variable integrality within BT loop.

This certainly shouldn't hold up this PR and should be added as an issue, if anything.

**Review by bknueven — APPROVED** [review]

File: coramin/domain_reduction/tests/test_decomp.py (hunk 1 above)
Comment (jsiirola) [reply coramin/domain_reduction/tests/test_decomp.py]: Agreed.

File: coramin/utils/pyomo_utils.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import pyomo.environ as pe
+
+
+def get_obj(m):
```
Comment (jsiirola) [inline coramin/utils/pyomo_utils.py]: This is a style thing, but I read this as "get object".  Would you be opposed to renaming it `get_objective()`?

**Review by jsiirola — COMMENTED** [review]
Some minor comments / questions, but nothing to hold up the PR.

**Comment by michaelbynum** [comment]
@bknueven @carldlaird @jsiirola Great suggestions everyone. Thank you all. Regarding the better inline documentation, I am working on readthedocs, including a tutorial section. I personally think this is a better place to work on usage docs. I couldn't decide where in `dbt.py` to put a high level description because several of the functions/classes are meant to be used together. Thoughts?

The readthedocs documentation will come in a separate PR.

**Pull request merged**